{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "471797c2-f594-4b4e-80ea-be4d0b70d7d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Experiment 4.1\n",
    "individual_sizes = [1000, 2000, 4000, 8000]\n",
    "x_labels = [\"High LD\", \"Middle LD\", \"Low LD\"]\n",
    "plot_relative_error_across_groups_combined(\n",
    "    highLD_largeDataset_allVariants,\n",
    "    middleLD_largeDataset_allVariants,\n",
    "    lowLD_largeDataset_allVariants,\n",
    "    x_labels=x_labels,\n",
    "    individual_sizes=individual_sizes,\n",
    "    col_num=0,\n",
    "    real_value=0.9,\n",
    "    ymin=-1,\n",
    "    ymax=1,\n",
    "    x_axis_name=\"LD Level\"\n",
    ")\n",
    "\n",
    "\n",
    "plot_relative_error_across_LD_levels"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "343db101-9b72-461d-b499-da3970cf4e63",
   "metadata": {},
   "source": [
    "#### **Experiment 4.2**:\n",
    "We then invetigate the influence of MAF.\n",
    "\n",
    "**Hypothesis**:  The MAF will influences the biasness of the epistatic heritability estimator $\\hat{h}^2_{g \\times g}$, The rarer the variant is, the more biased the estimation will be.\n",
    "**Simple Inference**: Still from the formule:\n",
    "\n",
    "$$\n",
    "\\text{Var}(Z_i Z_j) = 1 + \\frac{\\sigma_{ij}(1 - 2p_i)(1 - 2p_j)}{4p_i p_j (1 - p_i)(1 - p_j)}\n",
    "$$\n",
    "\n",
    "Since ld $r_{ij} = \\frac{\\sigma_{ij}}{\\sqrt{p_i p_j (1 - p_i)(1 - p_j)}}$:\n",
    "$$\n",
    "\\text{Var}(Z_i Z_j) = 1 + \\frac{r_{ij}(1 - 2p_i)(1 - 2p_j)}{4\\sqrt{p_i p_j (1 - p_i)(1 - p_j)}}\n",
    "$$\n",
    "\n",
    "let $f(p_i,p_j) =  \\frac{(1 - 2p_i)(1 - 2p_j)}{4\\sqrt{p_i p_j (1 - p_i)(1 - p_j)}}$\n",
    "$$\n",
    "\\text{Var}(Z_i Z_j) = 1 + r_{ij} f(p_i,p_j)\n",
    "$$\n",
    "\n",
    "Some property:\n",
    "1. $f(p_i,p_j) = 0$, when $p_i $ or $p_j $= 0.5\n",
    "\n",
    "2. $f(p_i,p_j) > 0$, when  both $p_i $ and $p_j $ on the same side( both > 0.5 or  both < 0.5) ; $f(p_i,p_j) < 0$, when they are on\n",
    "opposite sides\n",
    "\n",
    "3. $f(p_i,p_j) \\to âˆž$ when both allele frequencies are extreme (close to 0 or 1), rare variants (close to 0) dramatically inflate ($r_{ij} > 0 $) or deflate ($r_{ij} < 0 $) the $\\text{Var}(Z_i Z_j)$\n",
    "\n",
    "When two SNPs are independent ($r^2 =0, \\sigma_{ij} = 0$), the influence of MAF is diminishing, while the dataset is corrected, the low MAF will amplify the effect.\n",
    "\n",
    "\n",
    "**Test**:\n",
    "To examine the influence of MAF on estimation bias, we compared six datasets across three LD levels (high, middle, low), each with two MAF conditions (all variants vs. common variants only): highLD_large_all and highLD_large_common, middleLD_large_all and middleLD_large_common, lowLD_large_all and lowLD_large_common."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a5c6d4a-9ff0-42fc-ada9-c4bc558865bc",
   "metadata": {},
   "source": [
    "#### **Experiment 4.2 Summary**:\n",
    "1. Biasness: as we expected, when data is highly corrected, the bias is amplified (Figure 4.21 vs Figure 4.22), the effect dimishing with the reduce of LD level. However, the common variants groups show higher bias compared with all variants groups.\n",
    "\n",
    "2. Efficicy: Faster convergence for common variants, under same individual size common variants groups have lower SD (but take a look at Figure4.25 vs Figure4.26), This is reasonable.\n",
    "\n",
    "#### **Conclusion**:\n",
    "MAF influences both standard error and efficiency: restricting to common variants reduces variance and improves convergence speed. However, MAF also impacts bias, particularly under high LD conditions, where common variants groups unexpectedly showed greater bias than all variants groups.\n",
    "\n",
    "#### **Thinking**:\n",
    "Why do common variants exhibit higher bias than all variants? Is it possible that the common varaints shows higher LD since we use threshold to prune LD\n",
    "(The experiment may not be rigurous)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b69709fd-2448-4919-8fe5-754de1505b38",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
